validator/client/beacon-api/activation.go

```diff
@@ -0,0 +1,153 @@
+//go:build use_beacon_api
+// +build use_beacon_api
+
+package beacon_api
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/api/gateway/apimiddleware"
+	rpcapi "github.com/prysmaticlabs/prysm/v3/beacon-chain/rpc/apimiddleware"
+	"github.com/prysmaticlabs/prysm/v3/config/params"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"google.golang.org/grpc"
+)
+
+func (c beaconApiValidatorClient) waitForActivation(in *ethpb.ValidatorActivationRequest) (ethpb.BeaconNodeValidator_WaitForActivationClient, error) {
+	return &beaconNodeValidatorWaitForActivationClient{
+		beaconApiValidatorClient:   c,
+		ValidatorActivationRequest: in,
+		firstRecvCall:              true,
+	}, nil
+}
+
+type beaconNodeValidatorWaitForActivationClient struct {
+	grpc.ClientStream
+	beaconApiValidatorClient
+	*ethpb.ValidatorActivationRequest
+	firstRecvCall bool
+}
+
+func (x *beaconNodeValidatorWaitForActivationClient) Recv() (*ethpb.ValidatorActivationResponse, error) {
```

Comment:
My understanding is the code you quote:

https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/beacon-chain/rpc/prysm/v1alpha1/validator/server.go#L95-L118

is running on server side (Beacon Node) when using gRPC and not on client side (Validator Client).

> But how does the caller know when to stop looping?

The validator client actually runs a `for` loop here: 
https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/validator/client/wait_for_activation.go#L163-L212

It calls `Recv` on line 169.
- If no validator is active, it fails test on line 203 then goes to `continue` on line 206 (then goes back to the beginning of the loop).
- If at least one validator is active, it successes test on line 203 then goes to `break` line 209 then exits the `for` loop.

So, if the validator starts with no active key, it will continue to run the for loop.
Once at least once key is active, it will exit the for loop.

Note: Each call of `Recv` asks the BN for latest keys statuses.

-----------
With gRPC, there is actually 2 `for` loops running:
The first one this one, in the Beacon Node:
https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/beacon-chain/rpc/prysm/v1alpha1/validator/server.go#L95-L118

The second one is this one, in the Validator Client:
https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/validator/client/wait_for_activation.go#L163-L212


Using REST, there is no more `for` loop on Beacon Node.

==> I can incorporate into the `Recv` function  a `for+select+channel` with a `time.After` clause every 12 seconds which will loop and print some logs every slot until at least one validator is active, and which will return the list of statuses if one of them is active.

It will solve the "drift" problem, but will add some redundancy with the code here
https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/validator/client/wait_for_activation.go#L163-L212

which already loops until one validator is active.

(As I said earlier, I'm quite new at golang, so I may miss a very obvious point related to the language!)

Reply:
Let me come back with something more convincing :)

Reply:
I don't see the functional need to emulate on client side the for loop currently present on server side with gRPC.
I have the impression (at least, when I test it) that the behaviour is correct, except for the drift issue.
I can mitigate this drift issue by using a `for+select+channel` instead of the `wait`.